{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "413b998e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import collections\n",
    "from collections import OrderedDict, Counter, defaultdict\n",
    "import pandas as pd\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from bokeh.plotting import figure, output_file, show \n",
    "import matplotlib.patches as patches    \n",
    "from matplotlib import colors\n",
    "\n",
    "\n",
    "import glob\n",
    "\n",
    "import subprocess\n",
    "from subprocess import call\n",
    "\n",
    "import re\n",
    "\n",
    "import pickle\n",
    "\n",
    "#from Bio.SubsMat import MatrixInfo\n",
    "#import Bio.Data.CodonTable\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "import Bio.Data.CodonTable\n",
    "\n",
    "#import joypy\n",
    "\n",
    "import pyBigWig"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ae3775a",
   "metadata": {},
   "source": [
    "* Peptide: Sequence of the peptide\n",
    "* Gene_Names: The Genes that this peptide maps to. Per your suggestion, I've gotten rid of any peptides that matched to more than one gene. Also note that many peptides do not come from an ENSEMBL annotated gene and are thus left as '-'. \n",
    "* BestPSM_PEP: This is the Posterior Error Probability of the peptide. If you wanted to rank the peptides by some MS score, this is the best. \n",
    "* ENST_ids: All of the ENST ids that this peptide could come from. Notice that many of the peptide will come from the same gene but we are unable to distinguish between different transcripts. \n",
    "* Protein_Accessions: The accessions of the proteins it matched to. This is probably the least helpful column for you\n",
    "* Chromosome/scaffold name: Name of chromosome\n",
    "* Gene start (bp): Where we can map the peptide of the genome, this is where the gene starts \n",
    "* Gene end (bp): Where the gene ends\n",
    "* Parsed_Loci: For some peptides that do not map to ENSEMBL the original curators of the database have added genomic co-ordianates (which you can see in the protein_accessions column). This column just parses those locations. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "id": "dbbdbc89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Peptide</th>\n",
       "      <th>Gene_Names</th>\n",
       "      <th>BestPSM_PEP</th>\n",
       "      <th>ENST_ids</th>\n",
       "      <th>Protein_Accessions</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Parsed_Loci</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAPESF</td>\n",
       "      <td>CIZ1</td>\n",
       "      <td>0.010166</td>\n",
       "      <td>ENST00000476727</td>\n",
       "      <td>NonCoding_2_ENST00000476727_ENSG00000148337</td>\n",
       "      <td>9</td>\n",
       "      <td>128161251</td>\n",
       "      <td>128204383</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AADGHHRGL</td>\n",
       "      <td>HSPBP1</td>\n",
       "      <td>0.001552</td>\n",
       "      <td>ENST00000255631,ENST00000433386,ENST0000058569...</td>\n",
       "      <td>10_ENST00000255631_ENSG00000133265_F1_orf2,10_...</td>\n",
       "      <td>19</td>\n",
       "      <td>55262223</td>\n",
       "      <td>55280381</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AADHGPDEY</td>\n",
       "      <td>POMK</td>\n",
       "      <td>0.001915</td>\n",
       "      <td>ENST00000331373,ENST00000518991,ENST0000061442...</td>\n",
       "      <td>19_ENST00000331373_ENSG00000185900_F2_orf2,19_...</td>\n",
       "      <td>8</td>\n",
       "      <td>43093498</td>\n",
       "      <td>43131180</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAEPVDSSL</td>\n",
       "      <td>-</td>\n",
       "      <td>0.046727</td>\n",
       "      <td>-</td>\n",
       "      <td>ERV_1792085_ERVL-E-int::chr18:74213725-74215142_6</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr18:74213725-74215142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAFAAPVKA</td>\n",
       "      <td>DPY19L2</td>\n",
       "      <td>0.021480</td>\n",
       "      <td>ENST00000538147</td>\n",
       "      <td>2_ENST00000538147_ENSG00000177990_F0_orf2,UTR5...</td>\n",
       "      <td>12</td>\n",
       "      <td>63558913</td>\n",
       "      <td>63668939</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518</th>\n",
       "      <td>YTADHGHVY</td>\n",
       "      <td>CLDND1</td>\n",
       "      <td>0.000701</td>\n",
       "      <td>ENST00000503004</td>\n",
       "      <td>28_ENST00000503004_ENSG00000080822_F1_orf11</td>\n",
       "      <td>3</td>\n",
       "      <td>98497912</td>\n",
       "      <td>98523066</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1519</th>\n",
       "      <td>YTDHAHPLHRGDS</td>\n",
       "      <td>-</td>\n",
       "      <td>0.040256</td>\n",
       "      <td>-</td>\n",
       "      <td>ERV_1270162_MLT1K_MLT1K-int::chr15:29477473-29...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr15:29477473-29480097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520</th>\n",
       "      <td>YTDHSWLDQRW</td>\n",
       "      <td>-</td>\n",
       "      <td>0.024654</td>\n",
       "      <td>-</td>\n",
       "      <td>ERV_2415898_MLT1M::chr21:37752264-37752370_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr21:37752264-37752370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1521</th>\n",
       "      <td>YTLVHAGP</td>\n",
       "      <td>PSMF1</td>\n",
       "      <td>0.026704</td>\n",
       "      <td>ENST00000335877</td>\n",
       "      <td>153_ENST00000335877_ENSG00000125818_F2_orf81,U...</td>\n",
       "      <td>20</td>\n",
       "      <td>1113240</td>\n",
       "      <td>1189415</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1522</th>\n",
       "      <td>YVSLLDAPGHRD</td>\n",
       "      <td>-</td>\n",
       "      <td>0.025604</td>\n",
       "      <td>ENST00000325349</td>\n",
       "      <td>NonCoding_3_ENST00000325349_ENSG00000213235</td>\n",
       "      <td>12</td>\n",
       "      <td>16990670</td>\n",
       "      <td>16992083</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1523 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Peptide Gene_Names  BestPSM_PEP  \\\n",
       "0          AAAAPESF       CIZ1     0.010166   \n",
       "1         AADGHHRGL     HSPBP1     0.001552   \n",
       "2         AADHGPDEY       POMK     0.001915   \n",
       "3         AAEPVDSSL          -     0.046727   \n",
       "4         AAFAAPVKA    DPY19L2     0.021480   \n",
       "...             ...        ...          ...   \n",
       "1518      YTADHGHVY     CLDND1     0.000701   \n",
       "1519  YTDHAHPLHRGDS          -     0.040256   \n",
       "1520    YTDHSWLDQRW          -     0.024654   \n",
       "1521       YTLVHAGP      PSMF1     0.026704   \n",
       "1522   YVSLLDAPGHRD          -     0.025604   \n",
       "\n",
       "                                               ENST_ids  \\\n",
       "0                                       ENST00000476727   \n",
       "1     ENST00000255631,ENST00000433386,ENST0000058569...   \n",
       "2     ENST00000331373,ENST00000518991,ENST0000061442...   \n",
       "3                                                     -   \n",
       "4                                       ENST00000538147   \n",
       "...                                                 ...   \n",
       "1518                                    ENST00000503004   \n",
       "1519                                                  -   \n",
       "1520                                                  -   \n",
       "1521                                    ENST00000335877   \n",
       "1522                                    ENST00000325349   \n",
       "\n",
       "                                     Protein_Accessions  \\\n",
       "0           NonCoding_2_ENST00000476727_ENSG00000148337   \n",
       "1     10_ENST00000255631_ENSG00000133265_F1_orf2,10_...   \n",
       "2     19_ENST00000331373_ENSG00000185900_F2_orf2,19_...   \n",
       "3     ERV_1792085_ERVL-E-int::chr18:74213725-74215142_6   \n",
       "4     2_ENST00000538147_ENSG00000177990_F0_orf2,UTR5...   \n",
       "...                                                 ...   \n",
       "1518        28_ENST00000503004_ENSG00000080822_F1_orf11   \n",
       "1519  ERV_1270162_MLT1K_MLT1K-int::chr15:29477473-29...   \n",
       "1520       ERV_2415898_MLT1M::chr21:37752264-37752370_1   \n",
       "1521  153_ENST00000335877_ENSG00000125818_F2_orf81,U...   \n",
       "1522        NonCoding_3_ENST00000325349_ENSG00000213235   \n",
       "\n",
       "     Chromosome/scaffold name Gene start (bp) Gene end (bp)  \\\n",
       "0                           9       128161251     128204383   \n",
       "1                          19        55262223      55280381   \n",
       "2                           8        43093498      43131180   \n",
       "3                           -               -             -   \n",
       "4                          12        63558913      63668939   \n",
       "...                       ...             ...           ...   \n",
       "1518                        3        98497912      98523066   \n",
       "1519                        -               -             -   \n",
       "1520                        -               -             -   \n",
       "1521                       20         1113240       1189415   \n",
       "1522                       12        16990670      16992083   \n",
       "\n",
       "                  Parsed_Loci  \n",
       "0                           -  \n",
       "1                           -  \n",
       "2                           -  \n",
       "3     chr18:74213725-74215142  \n",
       "4                           -  \n",
       "...                       ...  \n",
       "1518                        -  \n",
       "1519  chr15:29477473-29480097  \n",
       "1520  chr21:37752264-37752370  \n",
       "1521                        -  \n",
       "1522                        -  \n",
       "\n",
       "[1523 rows x 9 columns]"
      ]
     },
     "execution_count": 345,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('DATA/Novels_forAlla.txt', sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "id": "b94f7828",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_tr</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>cds_seq</th>\n",
       "      <th>cds_start_codon</th>\n",
       "      <th>cds_stop_codon</th>\n",
       "      <th>cds_start_pos</th>\n",
       "      <th>cds_stop_pos</th>\n",
       "      <th>utr5_start</th>\n",
       "      <th>utr5_end</th>\n",
       "      <th>record_id</th>\n",
       "      <th>tr_id1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000335137.3</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATG</td>\n",
       "      <td>ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>0</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000335137.3|ENSG00000186092.4|OTTHUMG000...</td>\n",
       "      <td>ENST00000335137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>FO538757.2-201</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>AAG</td>\n",
       "      <td>ATGAGTGACAGCATCAACTTCTCTCACAACCTAGGCCAGCTCCTGT...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>316</td>\n",
       "      <td>718</td>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>ENST00000624431.1|ENSG00000279928.1|-|-|FO5387...</td>\n",
       "      <td>ENST00000624431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000623834.3</td>\n",
       "      <td>FO538757.1</td>\n",
       "      <td>FO538757.1-201</td>\n",
       "      <td>ATGACTCCTGTGAGGATGCAGCACTCCCTGGCAGGTCAGACCTATG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATG</td>\n",
       "      <td>ATGACTCCTGTGAGGATGCAGCACTCCCTGGCAGGTCAGACCTATG...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>0</td>\n",
       "      <td>1080</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000623834.3|ENSG00000279457.3|-|-|FO5387...</td>\n",
       "      <td>ENST00000623834</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id        gene         gene_tr  \\\n",
       "0  ENST00000335137.3       OR4F5       OR4F5-001   \n",
       "1  ENST00000624431.1  FO538757.2  FO538757.2-201   \n",
       "2  ENST00000623834.3  FO538757.1  FO538757.1-201   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...   \n",
       "1  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "2  ATGACTCCTGTGAGGATGCAGCACTCCCTGGCAGGTCAGACCTATG...   \n",
       "\n",
       "                                      5UTR_start_seq CDS_start  \\\n",
       "0                                                NaN       ATG   \n",
       "1  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...       AAG   \n",
       "2                                                NaN       ATG   \n",
       "\n",
       "                                             cds_seq cds_start_codon  \\\n",
       "0  ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...             ATG   \n",
       "1  ATGAGTGACAGCATCAACTTCTCTCACAACCTAGGCCAGCTCCTGT...             ATG   \n",
       "2  ATGACTCCTGTGAGGATGCAGCACTCCCTGGCAGGTCAGACCTATG...             ATG   \n",
       "\n",
       "  cds_stop_codon  cds_start_pos  cds_stop_pos  utr5_start  utr5_end  \\\n",
       "0            TAG              0           918           0         0   \n",
       "1            TAG            316           718           0       316   \n",
       "2            TAG              0          1080           0         0   \n",
       "\n",
       "                                           record_id           tr_id1  \n",
       "0  ENST00000335137.3|ENSG00000186092.4|OTTHUMG000...  ENST00000335137  \n",
       "1  ENST00000624431.1|ENSG00000279928.1|-|-|FO5387...  ENST00000624431  \n",
       "2  ENST00000623834.3|ENSG00000279457.3|-|-|FO5387...  ENST00000623834  "
      ]
     },
     "execution_count": 306,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_pc_g25 =  pd.read_csv('/home/alla/make_PhD_great_again/NTE/VERSION_since_May2021/tmp_res/metadata_pc_g25.txt', sep='\\t')\n",
    "metadata_pc_g25['tr_id1'] = [x.split('.')[0] for x in metadata_pc_g25['tr_id'].tolist()]\n",
    "metadata_pc_g25[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "id": "252e343c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = pd.read_csv('DATA/gencode.v25.annotation.gtf', sep='\\t', comment='#', header=None)\n",
    "gtf_tr = gtf[gtf[2] == 'transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "id": "eedb9571",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_74992/2591424268.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gtf_tr['tr_id'] = [x.split('transcript_id ')[1].split('\"')[1] for x in gtf_tr[8].tolist()]\n",
      "/tmp/ipykernel_74992/2591424268.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gtf_tr['tr_id1'] = [x.split('.')[0] for x in gtf_tr['tr_id'].tolist()]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>tr_id</th>\n",
       "      <th>tr_id1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENST00000456328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENST00000450305</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1           2      3      4  5  6  7  \\\n",
       "1  chr1  HAVANA  transcript  11869  14409  .  +  .   \n",
       "5  chr1  HAVANA  transcript  12010  13670  .  +  .   \n",
       "\n",
       "                                                   8              tr_id  \\\n",
       "1  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  ENST00000456328.2   \n",
       "5  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  ENST00000450305.2   \n",
       "\n",
       "            tr_id1  \n",
       "1  ENST00000456328  \n",
       "5  ENST00000450305  "
      ]
     },
     "execution_count": 308,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_tr['tr_id'] = [x.split('transcript_id ')[1].split('\"')[1] for x in gtf_tr[8].tolist()]\n",
    "gtf_tr['tr_id1'] = [x.split('.')[0] for x in gtf_tr['tr_id'].tolist()]\n",
    "\n",
    "gtf_tr[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "id": "ffe20cc5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'gene_id \"ENSG00000223972.5\"; transcript_id \"ENST00000456328.2\"; gene_type \"transcribed_unprocessed_pseudogene\"; gene_status \"KNOWN\"; gene_name \"DDX11L1\"; transcript_type \"processed_transcript\"; transcript_status \"KNOWN\"; transcript_name \"DDX11L1-002\"; level 2; transcript_support_level \"1\"; tag \"basic\"; havana_gene \"OTTHUMG00000000961.2\"; havana_transcript \"OTTHUMT00000362751.1\";'"
      ]
     },
     "execution_count": 274,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_tr.iloc[0][8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de6b86ca",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cd43a8c0",
   "metadata": {},
   "source": [
    "# We need to select reasonable rank for each category to make a cutoff "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 506,
   "id": "7b817a27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rank 5000\n",
    "uORF_df = pd.read_csv('DATA/NonCanonicalORFs_Trips-viz_10studies_100samples_1000/AUG_uORF_1000_10studies.csv')\n",
    "\n",
    "# rank y\n",
    "uoORF_df = pd.read_csv('DATA/NonCanonicalORFs_Trips-viz_10studies_100samples_1000/AUG_uoORF_1000_10studies.csv')\n",
    "\n",
    "# rank x\n",
    "doORF_df = pd.read_csv('DATA/NonCanonicalORFs_Trips-viz_10studies_100samples_1000/AUG_doORF_1000_10studies.csv')\n",
    "\n",
    "# rank u\n",
    "dORF_df = pd.read_csv('DATA/NonCanonicalORFs_Trips-viz_10studies_100samples_1000/AUG_dORF_1000_10studies.csv')\n",
    "\n",
    "# rank a\n",
    "lncORF_df = pd.read_csv('DATA/NonCanonicalORFs_Trips-viz_10studies_100samples_1000/AUG_lncORF_1000_10studies.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 507,
   "id": "c04b4f87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "145334 145334 145448 145240\n"
     ]
    }
   ],
   "source": [
    "# rank \n",
    "nORF_df = pd.read_csv('DATA/NonCanonicalORFs_Trips-viz_10studies_100samples_1000/AUG_nORF_1000_10studies.csv').drop_duplicates()\n",
    "nORF_df.columns = ['Gene', 'tr_id1', 'Start', 'Stop', 'Length', 'Global_Rank', 'Type',\n",
    "       'Trips-viz link', 'Start Codon', 'Highframe rank', 'Highframe value',\n",
    "       'Lowframe rank', 'Lowframe value', 'Stop rank', 'Stop value',\n",
    "       'Start rank', 'Start value', 'Coverage rank', 'Coverage value',\n",
    "       'Inframe Count Rank', 'Inframe Count Value', 'Amino acid sequence',\n",
    "       'Proteomics count', 'Read density']\n",
    "\n",
    "# filter reading frame of nORFs \n",
    "nORF_df[0:2]\n",
    "\n",
    "nORF_df1 = nORF_df.merge(metadata_pc_g25[['tr_id1', 'cds_start_pos']],\n",
    "                        on='tr_id1', how='inner').drop_duplicates()\n",
    "\n",
    "\n",
    "li = []\n",
    "for el in nORF_df1[['tr_id1', 'Start', 'cds_start_pos']].to_numpy():\n",
    "    tr_id1 = el[0]\n",
    "    start = el[1]-1\n",
    "    cds_start = el[2]\n",
    "    \n",
    "    if (cds_start % 3) == (start % 3):\n",
    "        # same frame \n",
    "        frame = 'the same'\n",
    "    else:\n",
    "        frame = 'different'\n",
    "        \n",
    "    li.append([tr_id1, el[1], frame])\n",
    "    \n",
    "frame_df = pd.DataFrame(li, columns = ['tr_id1', 'Start', 'frame'])\n",
    "    \n",
    "nORF_df2 = nORF_df1.merge(frame_df,\n",
    "                         on=['tr_id1', 'Start'], how='inner').drop_duplicates()\n",
    "\n",
    "\n",
    "# if frame is the same it's probably lncRNA or wrong CDS annotation with no 5'UTR\n",
    "nORF_df3 = nORF_df2[nORF_df2['frame'] == 'different']\n",
    "\n",
    "print (nORF_df2.shape[0], nORF_df1.shape[0], nORF_df.shape[0], nORF_df3.shape[0])\n",
    "\n",
    "# pseudogenes\n",
    "# set(nORF_df.tr_id1.tolist()) - set(nORF_df2.tr_id1.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6b3fdc6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79b8efde",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38ed020b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "437ff7d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "3b47fe74",
   "metadata": {},
   "source": [
    "# Combine all ORFs together using rank "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 495,
   "id": "ea3ed7b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "uORF_sel = uORF_df[['Gene', 'Tran', 'Start', 'Stop', 'Global_Rank', 'Type', \n",
    "         'Amino acid sequence']][0:6200]\n",
    "\n",
    "uORF_sel.columns = ['Gene', 'tr_id1', 'Start', \n",
    "                    'Stop', 'Global_Rank_within_ORF_type', \n",
    "                    'Type_of_ORF', 'Amino acid sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 496,
   "id": "1beab828",
   "metadata": {},
   "outputs": [],
   "source": [
    "nORF_sel = nORF_df3[['Gene', 'tr_id1', 'Start', 'Stop', 'Global_Rank', 'Type', \n",
    "         'Amino acid sequence']].sort_values(by='Global_Rank')[0:945]\n",
    "\n",
    "nORF_sel.columns = ['Gene', 'tr_id1', 'Start', \n",
    "                    'Stop', 'Global_Rank_within_ORF_type', \n",
    "                    'Type_of_ORF', 'Amino acid sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 497,
   "id": "2b0fb9fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "uoORF_sel = uoORF_df[['Gene', 'Tran', 'Start', 'Stop', 'Global_Rank', 'Type', \n",
    "         'Amino acid sequence']][0:800]\n",
    "\n",
    "uoORF_sel.columns = ['Gene', 'tr_id1', 'Start', \n",
    "                    'Stop', 'Global_Rank_within_ORF_type', \n",
    "                    'Type_of_ORF', 'Amino acid sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 498,
   "id": "ce8a2d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dORF_sel = dORF_df[['Gene', 'Tran', 'Start', 'Stop', 'Global_Rank', 'Type', \n",
    "         'Amino acid sequence']][0:350]\n",
    "\n",
    "dORF_sel.columns = ['Gene', 'tr_id1', 'Start', \n",
    "                    'Stop', 'Global_Rank_within_ORF_type', \n",
    "                    'Type_of_ORF', 'Amino acid sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "id": "48434723",
   "metadata": {},
   "outputs": [],
   "source": [
    "doORF_sel = doORF_df[['Gene', 'Tran', 'Start', 'Stop', 'Global_Rank', 'Type', \n",
    "         'Amino acid sequence']][0:110]\n",
    "\n",
    "doORF_sel.columns = ['Gene', 'tr_id1', 'Start', \n",
    "                    'Stop', 'Global_Rank_within_ORF_type', \n",
    "                    'Type_of_ORF', 'Amino acid sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 500,
   "id": "6f47b534",
   "metadata": {},
   "outputs": [],
   "source": [
    "lncORF_sel = lncORF_df[['Gene', 'Tran', 'Start', 'Stop', 'Global_Rank', 'Type', \n",
    "         'Amino acid sequence']][0:2800]\n",
    "\n",
    "lncORF_sel.columns = ['Gene', 'tr_id1', 'Start', \n",
    "                    'Stop', 'Global_Rank_within_ORF_type', \n",
    "                    'Type_of_ORF', 'Amino acid sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 501,
   "id": "3d3b71f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "comb = pd.concat([uORF_sel, uoORF_sel, \n",
    "                  nORF_sel, \n",
    "                  #lncORF_sel,\n",
    "                  doORF_sel, \n",
    "                  dORF_sel])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 509,
   "id": "83979fb3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8405"
      ]
     },
     "execution_count": 509,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 510,
   "id": "a1aa572b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8404\n"
     ]
    }
   ],
   "source": [
    "comb_df = comb.merge(metadata_pc_g25[['tr_id1', 'tr_id']], on='tr_id1', how='inner').drop_duplicates()\n",
    "\n",
    "comb_df = comb_df[~comb_df['tr_id'].str.contains('PAR')]\n",
    "\n",
    "print (comb_df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 511,
   "id": "47d37951",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2800, 2800)"
      ]
     },
     "execution_count": 511,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lncORF_sel2 = lncORF_sel.merge(gtf_tr[['tr_id', 'tr_id1']], on='tr_id1', how='inner')\n",
    "\n",
    "lncORF_sel.shape[0], lncORF_sel2.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 512,
   "id": "d20f059f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11204"
      ]
     },
     "execution_count": 512,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total = pd.concat([comb_df, lncORF_sel2], axis=0)\n",
    "\n",
    "total.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 513,
   "id": "aa27d71f",
   "metadata": {},
   "outputs": [],
   "source": [
    "total.to_csv('tmp_res/11204_ncORFs.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 514,
   "id": "724b714e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>tr_id1</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Global_Rank_within_ORF_type</th>\n",
       "      <th>Type_of_ORF</th>\n",
       "      <th>Amino acid sequence</th>\n",
       "      <th>tr_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNX17</td>\n",
       "      <td>ENST00000233575</td>\n",
       "      <td>118</td>\n",
       "      <td>147</td>\n",
       "      <td>1</td>\n",
       "      <td>uorf</td>\n",
       "      <td>MASEAAGTR</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNX17</td>\n",
       "      <td>ENST00000233575</td>\n",
       "      <td>48</td>\n",
       "      <td>128</td>\n",
       "      <td>2342</td>\n",
       "      <td>uorf</td>\n",
       "      <td>MPRRLRPRTDVPPPSHIGSQGSQNGE</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MRPL18</td>\n",
       "      <td>ENST00000367034</td>\n",
       "      <td>49</td>\n",
       "      <td>84</td>\n",
       "      <td>2</td>\n",
       "      <td>uorf</td>\n",
       "      <td>MAAPGERLSRP</td>\n",
       "      <td>ENST00000367034.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MRPL18</td>\n",
       "      <td>ENST00000367034</td>\n",
       "      <td>244</td>\n",
       "      <td>306</td>\n",
       "      <td>88</td>\n",
       "      <td>nested</td>\n",
       "      <td>MKLSPQNSPTGTPGTWSFYL</td>\n",
       "      <td>ENST00000367034.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCNK</td>\n",
       "      <td>ENST00000389879</td>\n",
       "      <td>18</td>\n",
       "      <td>50</td>\n",
       "      <td>3</td>\n",
       "      <td>uorf</td>\n",
       "      <td>MAAVGKERRR</td>\n",
       "      <td>ENST00000389879.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene           tr_id1  Start  Stop  Global_Rank_within_ORF_type  \\\n",
       "0   SNX17  ENST00000233575    118   147                            1   \n",
       "1   SNX17  ENST00000233575     48   128                         2342   \n",
       "2  MRPL18  ENST00000367034     49    84                            2   \n",
       "3  MRPL18  ENST00000367034    244   306                           88   \n",
       "4    CCNK  ENST00000389879     18    50                            3   \n",
       "\n",
       "  Type_of_ORF         Amino acid sequence              tr_id  \n",
       "0        uorf                   MASEAAGTR  ENST00000233575.6  \n",
       "1        uorf  MPRRLRPRTDVPPPSHIGSQGSQNGE  ENST00000233575.6  \n",
       "2        uorf                 MAAPGERLSRP  ENST00000367034.4  \n",
       "3      nested        MKLSPQNSPTGTPGTWSFYL  ENST00000367034.4  \n",
       "4        uorf                  MAAVGKERRR  ENST00000389879.9  "
      ]
     },
     "execution_count": 514,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7fd8ad6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e4f1fc91",
   "metadata": {},
   "source": [
    "# GENOMIC COORDINATES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 515,
   "id": "fb6b1079",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_global_coo(path_to_file_with_global_coo, colname):\n",
    "    global_coo_g25 = pd.read_csv(path_to_file_with_global_coo, sep='\\t')\n",
    "    global_coo_g25_f = global_coo_g25[global_coo_g25['hit'] == True]\n",
    "    global_coo_g25_f = global_coo_g25_f.sort_values(by=['seqnames', 'start'])\n",
    "    \n",
    "    global_coo_g25_f['global_coo'] = global_coo_g25_f['seqnames']+':'+global_coo_g25_f['start'].astype(str) +'-'+global_coo_g25_f['end'].astype(str)\n",
    "    global_coo_g25_f = global_coo_g25_f[['group', 'group_name', 'strand', 'global_coo']].groupby(['group', 'group_name', 'strand']).agg('+'.join).reset_index()\n",
    "    global_coo_g25_f.columns = ['group', 'tr_id', 'strand', colname]\n",
    "    return global_coo_g25_f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 517,
   "id": "90b7272e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>group_name</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_name</th>\n",
       "      <th>exon_rank</th>\n",
       "      <th>hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>27370639</td>\n",
       "      <td>27370668</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>64659</td>\n",
       "      <td>ENSE00000962997.3</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>27371269</td>\n",
       "      <td>27371268</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>64667</td>\n",
       "      <td>ENSE00003541824.1</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>27372623</td>\n",
       "      <td>27372622</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>64671</td>\n",
       "      <td>ENSE00003573225.1</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>27373247</td>\n",
       "      <td>27373246</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>64677</td>\n",
       "      <td>ENSE00003669336.1</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>27373861</td>\n",
       "      <td>27373860</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>64680</td>\n",
       "      <td>ENSE00003653420.1</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>27374085</td>\n",
       "      <td>27374084</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>64683</td>\n",
       "      <td>ENSE00003533996.1</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>27374346</td>\n",
       "      <td>27374345</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>64685</td>\n",
       "      <td>ENSE00003476021.1</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>27374689</td>\n",
       "      <td>27374688</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>64689</td>\n",
       "      <td>ENSE00003556896.1</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>27375061</td>\n",
       "      <td>27375060</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>64692</td>\n",
       "      <td>ENSE00003692961.1</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>chr2</td>\n",
       "      <td>27375506</td>\n",
       "      <td>27375505</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>64696</td>\n",
       "      <td>ENSE00003690576.1</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    group         group_name seqnames     start       end  width strand  \\\n",
       "1       1  ENST00000233575.6     chr2  27370639  27370668     30      +   \n",
       "2       1  ENST00000233575.6     chr2  27371269  27371268      0      +   \n",
       "3       1  ENST00000233575.6     chr2  27372623  27372622      0      +   \n",
       "4       1  ENST00000233575.6     chr2  27373247  27373246      0      +   \n",
       "5       1  ENST00000233575.6     chr2  27373861  27373860      0      +   \n",
       "6       1  ENST00000233575.6     chr2  27374085  27374084      0      +   \n",
       "7       1  ENST00000233575.6     chr2  27374346  27374345      0      +   \n",
       "8       1  ENST00000233575.6     chr2  27374689  27374688      0      +   \n",
       "9       1  ENST00000233575.6     chr2  27375061  27375060      0      +   \n",
       "10      1  ENST00000233575.6     chr2  27375506  27375505      0      +   \n",
       "\n",
       "    exon_id          exon_name  exon_rank    hit  \n",
       "1     64659  ENSE00000962997.3          1   True  \n",
       "2     64667  ENSE00003541824.1          2  False  \n",
       "3     64671  ENSE00003573225.1          3  False  \n",
       "4     64677  ENSE00003669336.1          4  False  \n",
       "5     64680  ENSE00003653420.1          5  False  \n",
       "6     64683  ENSE00003533996.1          6  False  \n",
       "7     64685  ENSE00003476021.1          7  False  \n",
       "8     64689  ENSE00003556896.1          8  False  \n",
       "9     64692  ENSE00003692961.1          9  False  \n",
       "10    64696  ENSE00003690576.1         10  False  "
      ]
     },
     "execution_count": 517,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coo = pd.read_csv('tmp_res/11204_ncORFs_global.txt', sep='\\t')\n",
    "\n",
    "coo[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 520,
   "id": "753de1c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11204\n"
     ]
    }
   ],
   "source": [
    "all_ORFs_coo = prepare_global_coo(path_to_file_with_global_coo='tmp_res/11294_ncORFs_global.txt',\n",
    "                                   colname = 'global_coo')\n",
    "\n",
    "print (all_ORFs_coo.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 522,
   "id": "82c96b4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11204"
      ]
     },
     "execution_count": 522,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total['group'] = list(range(1, total.shape[0]+1))\n",
    "\n",
    "total.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 523,
   "id": "4ca2106c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11204"
      ]
     },
     "execution_count": 523,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_global = total.merge(all_ORFs_coo, on=['tr_id', 'group'], how='inner')\n",
    "\n",
    "total_global.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 524,
   "id": "626ced92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>tr_id1</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Global_Rank_within_ORF_type</th>\n",
       "      <th>Type_of_ORF</th>\n",
       "      <th>Amino acid sequence</th>\n",
       "      <th>tr_id</th>\n",
       "      <th>group</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNX17</td>\n",
       "      <td>ENST00000233575</td>\n",
       "      <td>118</td>\n",
       "      <td>147</td>\n",
       "      <td>1</td>\n",
       "      <td>uorf</td>\n",
       "      <td>MASEAAGTR</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr2:27370639-27370668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNX17</td>\n",
       "      <td>ENST00000233575</td>\n",
       "      <td>48</td>\n",
       "      <td>128</td>\n",
       "      <td>2342</td>\n",
       "      <td>uorf</td>\n",
       "      <td>MPRRLRPRTDVPPPSHIGSQGSQNGE</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>chr2:27370569-27370649</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene           tr_id1  Start  Stop  Global_Rank_within_ORF_type  \\\n",
       "0  SNX17  ENST00000233575    118   147                            1   \n",
       "1  SNX17  ENST00000233575     48   128                         2342   \n",
       "\n",
       "  Type_of_ORF         Amino acid sequence              tr_id  group strand  \\\n",
       "0        uorf                   MASEAAGTR  ENST00000233575.6      1      +   \n",
       "1        uorf  MPRRLRPRTDVPPPSHIGSQGSQNGE  ENST00000233575.6      2      +   \n",
       "\n",
       "               global_coo  \n",
       "0  chr2:27370639-27370668  \n",
       "1  chr2:27370569-27370649  "
      ]
     },
     "execution_count": 524,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_global[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 525,
   "id": "e5f29a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_global.to_csv('tmp_res/11204_ncORFs_with_global_coo.txt', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 526,
   "id": "c7b8d2c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10997, 11072)"
      ]
     },
     "execution_count": 526,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_global['Amino acid sequence'].nunique(), total_global['global_coo'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 533,
   "id": "b30172c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>tr_id1</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Global_Rank_within_ORF_type</th>\n",
       "      <th>Type_of_ORF</th>\n",
       "      <th>Amino acid sequence</th>\n",
       "      <th>tr_id</th>\n",
       "      <th>group</th>\n",
       "      <th>strand</th>\n",
       "      <th>global_coo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNX17</td>\n",
       "      <td>ENST00000233575</td>\n",
       "      <td>118</td>\n",
       "      <td>147</td>\n",
       "      <td>1</td>\n",
       "      <td>uorf</td>\n",
       "      <td>MASEAAGTR</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr2:27370639-27370668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNX17</td>\n",
       "      <td>ENST00000233575</td>\n",
       "      <td>48</td>\n",
       "      <td>128</td>\n",
       "      <td>2342</td>\n",
       "      <td>uorf</td>\n",
       "      <td>MPRRLRPRTDVPPPSHIGSQGSQNGE</td>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>chr2:27370569-27370649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MRPL18</td>\n",
       "      <td>ENST00000367034</td>\n",
       "      <td>49</td>\n",
       "      <td>84</td>\n",
       "      <td>2</td>\n",
       "      <td>uorf</td>\n",
       "      <td>MAAPGERLSRP</td>\n",
       "      <td>ENST00000367034.4</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>chr6:159790514-159790549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MRPL18</td>\n",
       "      <td>ENST00000367034</td>\n",
       "      <td>244</td>\n",
       "      <td>306</td>\n",
       "      <td>88</td>\n",
       "      <td>nested</td>\n",
       "      <td>MKLSPQNSPTGTPGTWSFYL</td>\n",
       "      <td>ENST00000367034.4</td>\n",
       "      <td>4</td>\n",
       "      <td>+</td>\n",
       "      <td>chr6:159791009-159791071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCNK</td>\n",
       "      <td>ENST00000389879</td>\n",
       "      <td>18</td>\n",
       "      <td>50</td>\n",
       "      <td>3</td>\n",
       "      <td>uorf</td>\n",
       "      <td>MAAVGKERRR</td>\n",
       "      <td>ENST00000389879.9</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "      <td>chr14:99481426-99481458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11199</th>\n",
       "      <td>CTD-2341M24.1</td>\n",
       "      <td>ENST00000553668</td>\n",
       "      <td>157</td>\n",
       "      <td>219</td>\n",
       "      <td>2794</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>MTFLLFAFGLASFIFLIHRG</td>\n",
       "      <td>ENST00000553668.2</td>\n",
       "      <td>11200</td>\n",
       "      <td>+</td>\n",
       "      <td>chr14:86070942-86070967+chr14:86127950-86127986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11200</th>\n",
       "      <td>CTC-351M12.1</td>\n",
       "      <td>ENST00000624833</td>\n",
       "      <td>206</td>\n",
       "      <td>286</td>\n",
       "      <td>2795</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>MSIHLGTFPSSAFEDGCARQAVKDKH</td>\n",
       "      <td>ENST00000624833.1</td>\n",
       "      <td>11201</td>\n",
       "      <td>-</td>\n",
       "      <td>chr5:69189167-69189247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11201</th>\n",
       "      <td>RP11-1070N10.5</td>\n",
       "      <td>ENST00000555032</td>\n",
       "      <td>121</td>\n",
       "      <td>177</td>\n",
       "      <td>2798</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>MDKLRAQWVMTCPESWRE</td>\n",
       "      <td>ENST00000555032.1</td>\n",
       "      <td>11202</td>\n",
       "      <td>+</td>\n",
       "      <td>chr14:95639953-95640009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11202</th>\n",
       "      <td>CTC-575D19.1</td>\n",
       "      <td>ENST00000522551</td>\n",
       "      <td>226</td>\n",
       "      <td>645</td>\n",
       "      <td>2799</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>MVKSCGKDGFHIRVRLHSFHVIRINKMLSCAGADRLQTGMRGAFGK...</td>\n",
       "      <td>ENST00000522551.1</td>\n",
       "      <td>11203</td>\n",
       "      <td>+</td>\n",
       "      <td>chr5:168616577-168616996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11203</th>\n",
       "      <td>RNA5SP161</td>\n",
       "      <td>ENST00000363871</td>\n",
       "      <td>23</td>\n",
       "      <td>55</td>\n",
       "      <td>2800</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>MRPISSDLGS</td>\n",
       "      <td>ENST00000363871.1</td>\n",
       "      <td>11204</td>\n",
       "      <td>+</td>\n",
       "      <td>chr4:56097412-56097444</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11204 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene           tr_id1 Start Stop Global_Rank_within_ORF_type  \\\n",
       "0               SNX17  ENST00000233575   118  147                           1   \n",
       "1               SNX17  ENST00000233575    48  128                        2342   \n",
       "2              MRPL18  ENST00000367034    49   84                           2   \n",
       "3              MRPL18  ENST00000367034   244  306                          88   \n",
       "4                CCNK  ENST00000389879    18   50                           3   \n",
       "...               ...              ...   ...  ...                         ...   \n",
       "11199   CTD-2341M24.1  ENST00000553668   157  219                        2794   \n",
       "11200    CTC-351M12.1  ENST00000624833   206  286                        2795   \n",
       "11201  RP11-1070N10.5  ENST00000555032   121  177                        2798   \n",
       "11202    CTC-575D19.1  ENST00000522551   226  645                        2799   \n",
       "11203       RNA5SP161  ENST00000363871    23   55                        2800   \n",
       "\n",
       "      Type_of_ORF                                Amino acid sequence  \\\n",
       "0            uorf                                          MASEAAGTR   \n",
       "1            uorf                         MPRRLRPRTDVPPPSHIGSQGSQNGE   \n",
       "2            uorf                                        MAAPGERLSRP   \n",
       "3          nested                               MKLSPQNSPTGTPGTWSFYL   \n",
       "4            uorf                                         MAAVGKERRR   \n",
       "...           ...                                                ...   \n",
       "11199   noncoding                               MTFLLFAFGLASFIFLIHRG   \n",
       "11200   noncoding                         MSIHLGTFPSSAFEDGCARQAVKDKH   \n",
       "11201   noncoding                                 MDKLRAQWVMTCPESWRE   \n",
       "11202   noncoding  MVKSCGKDGFHIRVRLHSFHVIRINKMLSCAGADRLQTGMRGAFGK...   \n",
       "11203   noncoding                                         MRPISSDLGS   \n",
       "\n",
       "                   tr_id  group strand  \\\n",
       "0      ENST00000233575.6      1      +   \n",
       "1      ENST00000233575.6      2      +   \n",
       "2      ENST00000367034.4      3      +   \n",
       "3      ENST00000367034.4      4      +   \n",
       "4      ENST00000389879.9      5      +   \n",
       "...                  ...    ...    ...   \n",
       "11199  ENST00000553668.2  11200      +   \n",
       "11200  ENST00000624833.1  11201      -   \n",
       "11201  ENST00000555032.1  11202      +   \n",
       "11202  ENST00000522551.1  11203      +   \n",
       "11203  ENST00000363871.1  11204      +   \n",
       "\n",
       "                                            global_coo  \n",
       "0                               chr2:27370639-27370668  \n",
       "1                               chr2:27370569-27370649  \n",
       "2                             chr6:159790514-159790549  \n",
       "3                             chr6:159791009-159791071  \n",
       "4                              chr14:99481426-99481458  \n",
       "...                                                ...  \n",
       "11199  chr14:86070942-86070967+chr14:86127950-86127986  \n",
       "11200                           chr5:69189167-69189247  \n",
       "11201                          chr14:95639953-95640009  \n",
       "11202                         chr5:168616577-168616996  \n",
       "11203                           chr4:56097412-56097444  \n",
       "\n",
       "[11204 rows x 11 columns]"
      ]
     },
     "execution_count": 533,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_global"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52dbc954",
   "metadata": {},
   "source": [
    "# Group ncORFs based on genomic coordinates (e.g. [nORF,uORF], [tr_id1,tr_id2]) -> how many ncORFs unique by genomic coo and aa?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 527,
   "id": "d40fe6aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11204"
      ]
     },
     "execution_count": 527,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_global = total.merge(all_ORFs_coo, on=['tr_id', 'group'], how='inner')\n",
    "\n",
    "total_global['Start'] = total_global['Start'].astype(str)\n",
    "\n",
    "total_global['Stop'] = total_global['Stop'].astype(str)\n",
    "\n",
    "total_global['Global_Rank_within_ORF_type'] = total_global['Global_Rank_within_ORF_type'].astype(str)\n",
    "\n",
    "total_global.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 528,
   "id": "0d1ced02",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_global['tr_id'] = total_global.groupby(['global_coo'])['tr_id'].transform(lambda x : ','.join(x))\n",
    "\n",
    "total_global['Type_of_ORF'] = total_global.groupby(['global_coo'])['Type_of_ORF'].transform(lambda x : ','.join(x))\n",
    "\n",
    "total_global['Start'] = total_global.groupby(['global_coo'])['Start'].transform(lambda x : ','.join(x))\n",
    "\n",
    "total_global['Stop'] = total_global.groupby(['global_coo'])['Stop'].transform(lambda x : ','.join(x))\n",
    "\n",
    "#total_global['Start'] = total_global.groupby(['global_coo'])['Start'].transform(lambda x : ','.join(x))\n",
    "\n",
    "total_global['Global_Rank_within_ORF_type'] = total_global.groupby(['global_coo'])['Global_Rank_within_ORF_type'].transform(lambda x : ','.join(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 531,
   "id": "28dda379",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Type_of_ORF</th>\n",
       "      <th>Global_Rank_within_ORF_type</th>\n",
       "      <th>strand</th>\n",
       "      <th>Amino acid sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>global_coo</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10:100185661-100185771</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:172711092-172711148</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:172750699-172750743</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:172750733-172750807</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:172750792-172750824</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr6:31666887-31666903+chr6:31667868-31667946</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11:65423982-65424071</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr17:42130406-42130471</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11:5345810-5345881</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr14:20684620-20684682</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11072 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               tr_id  Gene  Start  Stop  \\\n",
       "global_coo                                                                \n",
       "chr10:100185661-100185771                          1     1      1     1   \n",
       "chr3:172711092-172711148                           1     1      1     1   \n",
       "chr3:172750699-172750743                           1     1      1     1   \n",
       "chr3:172750733-172750807                           1     1      1     1   \n",
       "chr3:172750792-172750824                           1     1      1     1   \n",
       "...                                              ...   ...    ...   ...   \n",
       "chr6:31666887-31666903+chr6:31667868-31667946      2     2      2     2   \n",
       "chr11:65423982-65424071                            2     2      2     2   \n",
       "chr17:42130406-42130471                            2     2      2     2   \n",
       "chr11:5345810-5345881                              2     2      2     2   \n",
       "chr14:20684620-20684682                            3     3      3     3   \n",
       "\n",
       "                                               Type_of_ORF  \\\n",
       "global_coo                                                   \n",
       "chr10:100185661-100185771                                1   \n",
       "chr3:172711092-172711148                                 1   \n",
       "chr3:172750699-172750743                                 1   \n",
       "chr3:172750733-172750807                                 1   \n",
       "chr3:172750792-172750824                                 1   \n",
       "...                                                    ...   \n",
       "chr6:31666887-31666903+chr6:31667868-31667946            2   \n",
       "chr11:65423982-65424071                                  2   \n",
       "chr17:42130406-42130471                                  2   \n",
       "chr11:5345810-5345881                                    2   \n",
       "chr14:20684620-20684682                                  3   \n",
       "\n",
       "                                               Global_Rank_within_ORF_type  \\\n",
       "global_coo                                                                   \n",
       "chr10:100185661-100185771                                                1   \n",
       "chr3:172711092-172711148                                                 1   \n",
       "chr3:172750699-172750743                                                 1   \n",
       "chr3:172750733-172750807                                                 1   \n",
       "chr3:172750792-172750824                                                 1   \n",
       "...                                                                    ...   \n",
       "chr6:31666887-31666903+chr6:31667868-31667946                            2   \n",
       "chr11:65423982-65424071                                                  2   \n",
       "chr17:42130406-42130471                                                  2   \n",
       "chr11:5345810-5345881                                                    2   \n",
       "chr14:20684620-20684682                                                  3   \n",
       "\n",
       "                                               strand  Amino acid sequence  \n",
       "global_coo                                                                  \n",
       "chr10:100185661-100185771                           1                    1  \n",
       "chr3:172711092-172711148                            1                    1  \n",
       "chr3:172750699-172750743                            1                    1  \n",
       "chr3:172750733-172750807                            1                    1  \n",
       "chr3:172750792-172750824                            1                    1  \n",
       "...                                               ...                  ...  \n",
       "chr6:31666887-31666903+chr6:31667868-31667946       2                    2  \n",
       "chr11:65423982-65424071                             2                    2  \n",
       "chr17:42130406-42130471                             2                    2  \n",
       "chr11:5345810-5345881                               2                    2  \n",
       "chr14:20684620-20684682                             3                    3  \n",
       "\n",
       "[11072 rows x 8 columns]"
      ]
     },
     "execution_count": 531,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_global_agg = total_global[['tr_id', 'Gene', 'Start', 'Stop', 'Type_of_ORF',\n",
    "                                'Global_Rank_within_ORF_type', 'global_coo', 'strand', \n",
    "                                'Amino acid sequence']].drop_duplicates()\n",
    "\n",
    "total_global_agg.groupby(by='global_coo').count().sort_values(by='tr_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 530,
   "id": "c499b046",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Type_of_ORF</th>\n",
       "      <th>Global_Rank_within_ORF_type</th>\n",
       "      <th>global_coo</th>\n",
       "      <th>strand</th>\n",
       "      <th>Amino acid sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1534</th>\n",
       "      <td>ENST00000336811.10,ENST00000555835.2,ENST00000...</td>\n",
       "      <td>ANG</td>\n",
       "      <td>444,521,34</td>\n",
       "      <td>506,583,96</td>\n",
       "      <td>uorf,uorf,uorf</td>\n",
       "      <td>828,829,832</td>\n",
       "      <td>chr14:20684620-20684682</td>\n",
       "      <td>+</td>\n",
       "      <td>MQSRIQKNIFAFSTGSPSSY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536</th>\n",
       "      <td>ENST00000336811.10,ENST00000555835.2,ENST00000...</td>\n",
       "      <td>RNASE4</td>\n",
       "      <td>444,521,34</td>\n",
       "      <td>506,583,96</td>\n",
       "      <td>uorf,uorf,uorf</td>\n",
       "      <td>828,829,832</td>\n",
       "      <td>chr14:20684620-20684682</td>\n",
       "      <td>+</td>\n",
       "      <td>MQSRIQKNIFAFSTGSPSSY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1540</th>\n",
       "      <td>ENST00000336811.10,ENST00000555835.2,ENST00000...</td>\n",
       "      <td>RP11-903H12.5</td>\n",
       "      <td>444,521,34</td>\n",
       "      <td>506,583,96</td>\n",
       "      <td>uorf,uorf,uorf</td>\n",
       "      <td>828,829,832</td>\n",
       "      <td>chr14:20684620-20684682</td>\n",
       "      <td>+</td>\n",
       "      <td>MQSRIQKNIFAFSTGSPSSY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  tr_id           Gene  \\\n",
       "1534  ENST00000336811.10,ENST00000555835.2,ENST00000...            ANG   \n",
       "1536  ENST00000336811.10,ENST00000555835.2,ENST00000...         RNASE4   \n",
       "1540  ENST00000336811.10,ENST00000555835.2,ENST00000...  RP11-903H12.5   \n",
       "\n",
       "           Start        Stop     Type_of_ORF Global_Rank_within_ORF_type  \\\n",
       "1534  444,521,34  506,583,96  uorf,uorf,uorf                 828,829,832   \n",
       "1536  444,521,34  506,583,96  uorf,uorf,uorf                 828,829,832   \n",
       "1540  444,521,34  506,583,96  uorf,uorf,uorf                 828,829,832   \n",
       "\n",
       "                   global_coo strand   Amino acid sequence  \n",
       "1534  chr14:20684620-20684682      +  MQSRIQKNIFAFSTGSPSSY  \n",
       "1536  chr14:20684620-20684682      +  MQSRIQKNIFAFSTGSPSSY  \n",
       "1540  chr14:20684620-20684682      +  MQSRIQKNIFAFSTGSPSSY  "
      ]
     },
     "execution_count": 530,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_global_agg[total_global_agg['global_coo'] == 'chr14:20684620-20684682']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "97252ce1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Type_of_ORF</th>\n",
       "      <th>Global_Rank_within_ORF_type</th>\n",
       "      <th>global_coo</th>\n",
       "      <th>strand</th>\n",
       "      <th>Amino acid sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>SNX17</td>\n",
       "      <td>118</td>\n",
       "      <td>147</td>\n",
       "      <td>uorf</td>\n",
       "      <td>1</td>\n",
       "      <td>chr2:27370639-27370668</td>\n",
       "      <td>+</td>\n",
       "      <td>MASEAAGTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000233575.6</td>\n",
       "      <td>SNX17</td>\n",
       "      <td>48</td>\n",
       "      <td>128</td>\n",
       "      <td>uorf</td>\n",
       "      <td>2342</td>\n",
       "      <td>chr2:27370569-27370649</td>\n",
       "      <td>+</td>\n",
       "      <td>MPRRLRPRTDVPPPSHIGSQGSQNGE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000367034.4</td>\n",
       "      <td>MRPL18</td>\n",
       "      <td>49</td>\n",
       "      <td>84</td>\n",
       "      <td>uorf</td>\n",
       "      <td>2</td>\n",
       "      <td>chr6:159790514-159790549</td>\n",
       "      <td>+</td>\n",
       "      <td>MAAPGERLSRP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000389879.9</td>\n",
       "      <td>CCNK</td>\n",
       "      <td>18</td>\n",
       "      <td>50</td>\n",
       "      <td>uorf</td>\n",
       "      <td>3</td>\n",
       "      <td>chr14:99481426-99481458</td>\n",
       "      <td>+</td>\n",
       "      <td>MAAVGKERRR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000360150.8</td>\n",
       "      <td>ERGIC2</td>\n",
       "      <td>11</td>\n",
       "      <td>52</td>\n",
       "      <td>uorf</td>\n",
       "      <td>4</td>\n",
       "      <td>chr12:29371658-29371670+chr12:29381115-29381143</td>\n",
       "      <td>-</td>\n",
       "      <td>MAVGWDHNTSMTI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8295</th>\n",
       "      <td>ENST00000424112.2</td>\n",
       "      <td>HDAC11-AS1</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>1992</td>\n",
       "      <td>chr3:13479990-13480022</td>\n",
       "      <td>-</td>\n",
       "      <td>MSCSCASQAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8296</th>\n",
       "      <td>ENST00000416928.6</td>\n",
       "      <td>HAGLR</td>\n",
       "      <td>610</td>\n",
       "      <td>645</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>1993</td>\n",
       "      <td>chr2:176177424-176177459</td>\n",
       "      <td>-</td>\n",
       "      <td>MIGYSSDLALP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8297</th>\n",
       "      <td>ENST00000427161.1</td>\n",
       "      <td>RP11-87N24.3</td>\n",
       "      <td>89</td>\n",
       "      <td>139</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>1995</td>\n",
       "      <td>chr9:10631768-10631785+chr9:10632083-10632115</td>\n",
       "      <td>-</td>\n",
       "      <td>MVFFFFFFFFFFFFLC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8298</th>\n",
       "      <td>ENST00000609661.5</td>\n",
       "      <td>LINC00452</td>\n",
       "      <td>726</td>\n",
       "      <td>899</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>1999</td>\n",
       "      <td>chr13:113919549-113919654+chr13:113920723-1139...</td>\n",
       "      <td>+</td>\n",
       "      <td>MPSTPSEFPAAALGPAQRAWAQHLETPKRQTGSCPGFVRSLPRLPS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8299</th>\n",
       "      <td>ENST00000624270.1</td>\n",
       "      <td>RP11-334C17.3</td>\n",
       "      <td>114</td>\n",
       "      <td>575</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>2000</td>\n",
       "      <td>chr17:80147363-80147824</td>\n",
       "      <td>+</td>\n",
       "      <td>MEGVVLLKLAVFRTGRGISRTGALRFFERRVDLNSLETGWGREGPP...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8300 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  tr_id           Gene Start Stop Type_of_ORF  \\\n",
       "0     ENST00000233575.6          SNX17   118  147        uorf   \n",
       "1     ENST00000233575.6          SNX17    48  128        uorf   \n",
       "2     ENST00000367034.4         MRPL18    49   84        uorf   \n",
       "3     ENST00000389879.9           CCNK    18   50        uorf   \n",
       "4     ENST00000360150.8         ERGIC2    11   52        uorf   \n",
       "...                 ...            ...   ...  ...         ...   \n",
       "8295  ENST00000424112.2     HDAC11-AS1    32   64   noncoding   \n",
       "8296  ENST00000416928.6          HAGLR   610  645   noncoding   \n",
       "8297  ENST00000427161.1   RP11-87N24.3    89  139   noncoding   \n",
       "8298  ENST00000609661.5      LINC00452   726  899   noncoding   \n",
       "8299  ENST00000624270.1  RP11-334C17.3   114  575   noncoding   \n",
       "\n",
       "     Global_Rank_within_ORF_type  \\\n",
       "0                              1   \n",
       "1                           2342   \n",
       "2                              2   \n",
       "3                              3   \n",
       "4                              4   \n",
       "...                          ...   \n",
       "8295                        1992   \n",
       "8296                        1993   \n",
       "8297                        1995   \n",
       "8298                        1999   \n",
       "8299                        2000   \n",
       "\n",
       "                                             global_coo strand  \\\n",
       "0                                chr2:27370639-27370668      +   \n",
       "1                                chr2:27370569-27370649      +   \n",
       "2                              chr6:159790514-159790549      +   \n",
       "3                               chr14:99481426-99481458      +   \n",
       "4       chr12:29371658-29371670+chr12:29381115-29381143      -   \n",
       "...                                                 ...    ...   \n",
       "8295                             chr3:13479990-13480022      -   \n",
       "8296                           chr2:176177424-176177459      -   \n",
       "8297      chr9:10631768-10631785+chr9:10632083-10632115      -   \n",
       "8298  chr13:113919549-113919654+chr13:113920723-1139...      +   \n",
       "8299                            chr17:80147363-80147824      +   \n",
       "\n",
       "                                    Amino acid sequence  \n",
       "0                                             MASEAAGTR  \n",
       "1                            MPRRLRPRTDVPPPSHIGSQGSQNGE  \n",
       "2                                           MAAPGERLSRP  \n",
       "3                                            MAAVGKERRR  \n",
       "4                                         MAVGWDHNTSMTI  \n",
       "...                                                 ...  \n",
       "8295                                         MSCSCASQAG  \n",
       "8296                                        MIGYSSDLALP  \n",
       "8297                                   MVFFFFFFFFFFFFLC  \n",
       "8298  MPSTPSEFPAAALGPAQRAWAQHLETPKRQTGSCPGFVRSLPRLPS...  \n",
       "8299  MEGVVLLKLAVFRTGRGISRTGALRFFERRVDLNSLETGWGREGPP...  \n",
       "\n",
       "[8300 rows x 9 columns]"
      ]
     },
     "execution_count": 339,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_global_agg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fda0a51",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4fb07089",
   "metadata": {},
   "source": [
    "# Overlap with HLA-peptidomics data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 534,
   "id": "f4b88e9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10997\n"
     ]
    }
   ],
   "source": [
    "# only unique aa sequences \n",
    "\n",
    "unique_aa = list(total_global['Amino acid sequence'].unique())\n",
    "\n",
    "print (len(unique_aa))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 535,
   "id": "50cd6da3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11204"
      ]
     },
     "execution_count": 535,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_global.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 536,
   "id": "49df8eea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Peptide</th>\n",
       "      <th>Gene_Names</th>\n",
       "      <th>BestPSM_PEP</th>\n",
       "      <th>ENST_ids</th>\n",
       "      <th>Protein_Accessions</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Parsed_Loci</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAPESF</td>\n",
       "      <td>CIZ1</td>\n",
       "      <td>0.010166</td>\n",
       "      <td>ENST00000476727</td>\n",
       "      <td>NonCoding_2_ENST00000476727_ENSG00000148337</td>\n",
       "      <td>9</td>\n",
       "      <td>128161251</td>\n",
       "      <td>128204383</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AADGHHRGL</td>\n",
       "      <td>HSPBP1</td>\n",
       "      <td>0.001552</td>\n",
       "      <td>ENST00000255631,ENST00000433386,ENST0000058569...</td>\n",
       "      <td>10_ENST00000255631_ENSG00000133265_F1_orf2,10_...</td>\n",
       "      <td>19</td>\n",
       "      <td>55262223</td>\n",
       "      <td>55280381</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AADHGPDEY</td>\n",
       "      <td>POMK</td>\n",
       "      <td>0.001915</td>\n",
       "      <td>ENST00000331373,ENST00000518991,ENST0000061442...</td>\n",
       "      <td>19_ENST00000331373_ENSG00000185900_F2_orf2,19_...</td>\n",
       "      <td>8</td>\n",
       "      <td>43093498</td>\n",
       "      <td>43131180</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAEPVDSSL</td>\n",
       "      <td>-</td>\n",
       "      <td>0.046727</td>\n",
       "      <td>-</td>\n",
       "      <td>ERV_1792085_ERVL-E-int::chr18:74213725-74215142_6</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chr18:74213725-74215142</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Peptide Gene_Names  BestPSM_PEP  \\\n",
       "0   AAAAPESF       CIZ1     0.010166   \n",
       "1  AADGHHRGL     HSPBP1     0.001552   \n",
       "2  AADHGPDEY       POMK     0.001915   \n",
       "3  AAEPVDSSL          -     0.046727   \n",
       "\n",
       "                                            ENST_ids  \\\n",
       "0                                    ENST00000476727   \n",
       "1  ENST00000255631,ENST00000433386,ENST0000058569...   \n",
       "2  ENST00000331373,ENST00000518991,ENST0000061442...   \n",
       "3                                                  -   \n",
       "\n",
       "                                  Protein_Accessions Chromosome/scaffold name  \\\n",
       "0        NonCoding_2_ENST00000476727_ENSG00000148337                        9   \n",
       "1  10_ENST00000255631_ENSG00000133265_F1_orf2,10_...                       19   \n",
       "2  19_ENST00000331373_ENSG00000185900_F2_orf2,19_...                        8   \n",
       "3  ERV_1792085_ERVL-E-int::chr18:74213725-74215142_6                        -   \n",
       "\n",
       "  Gene start (bp) Gene end (bp)              Parsed_Loci  \n",
       "0       128161251     128204383                        -  \n",
       "1        55262223      55280381                        -  \n",
       "2        43093498      43131180                        -  \n",
       "3               -             -  chr18:74213725-74215142  "
      ]
     },
     "execution_count": 536,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('DATA/Novels_forAlla.txt', sep='\\t')\n",
    "df[0:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 537,
   "id": "692a1109",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0             NonCoding_2_ENST00000476727_ENSG00000148337\n",
       "1       10_ENST00000255631_ENSG00000133265_F1_orf2,10_...\n",
       "2       19_ENST00000331373_ENSG00000185900_F2_orf2,19_...\n",
       "3       ERV_1792085_ERVL-E-int::chr18:74213725-74215142_6\n",
       "4       2_ENST00000538147_ENSG00000177990_F0_orf2,UTR5...\n",
       "                              ...                        \n",
       "1518          28_ENST00000503004_ENSG00000080822_F1_orf11\n",
       "1519    ERV_1270162_MLT1K_MLT1K-int::chr15:29477473-29...\n",
       "1520         ERV_2415898_MLT1M::chr21:37752264-37752370_1\n",
       "1521    153_ENST00000335877_ENSG00000125818_F2_orf81,U...\n",
       "1522          NonCoding_3_ENST00000325349_ENSG00000213235\n",
       "Name: Protein_Accessions, Length: 1523, dtype: object"
      ]
     },
     "execution_count": 537,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Protein_Accessions']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 538,
   "id": "d6bf3337",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "2000\n",
      "2100\n",
      "2200\n",
      "2300\n",
      "2400\n",
      "2500\n",
      "2600\n",
      "2700\n",
      "2800\n",
      "2900\n",
      "3000\n",
      "3100\n",
      "3200\n",
      "3300\n",
      "3400\n",
      "3500\n",
      "3600\n",
      "3700\n",
      "3800\n",
      "3900\n",
      "4000\n",
      "4100\n",
      "4200\n",
      "4300\n",
      "4400\n",
      "4500\n",
      "4600\n",
      "4700\n",
      "4800\n",
      "4900\n",
      "5000\n",
      "5100\n",
      "5200\n",
      "5300\n",
      "5400\n",
      "5500\n",
      "5600\n",
      "5700\n",
      "5800\n",
      "5900\n",
      "6000\n",
      "6100\n",
      "6200\n",
      "6300\n",
      "6400\n",
      "6500\n",
      "6600\n",
      "6700\n",
      "6800\n",
      "6900\n",
      "7000\n",
      "7100\n",
      "7200\n",
      "7300\n",
      "7400\n",
      "7500\n",
      "7600\n",
      "7700\n",
      "7800\n",
      "7900\n",
      "8000\n",
      "8100\n",
      "8200\n",
      "8300\n",
      "8400\n",
      "8500\n",
      "8600\n",
      "8700\n",
      "8800\n",
      "8900\n",
      "9000\n",
      "9100\n",
      "9200\n",
      "9300\n",
      "9400\n",
      "9500\n",
      "9600\n",
      "9700\n",
      "9800\n",
      "9900\n",
      "10000\n",
      "10100\n",
      "10200\n",
      "10300\n",
      "10400\n",
      "10500\n",
      "10600\n",
      "10700\n",
      "10800\n",
      "10900\n"
     ]
    }
   ],
   "source": [
    "li = []\n",
    "\n",
    "for i, aa in enumerate(unique_aa):\n",
    "    for peptide in df['Peptide'].tolist():\n",
    "        if aa.find(peptide) != -1:\n",
    "            li.append([aa, peptide, aa.find(peptide)])\n",
    "    if i % 100 == 0:\n",
    "        print (i)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 547,
   "id": "c024fa74",
   "metadata": {},
   "outputs": [],
   "source": [
    "outdf = pd.DataFrame(li, columns = ['Amino acid sequence', 'peptide', 'pos'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 549,
   "id": "db924a08",
   "metadata": {},
   "outputs": [],
   "source": [
    "found_ORFs = outdf.merge(total_global_agg, \n",
    "                         on='Amino acid sequence', how='inner').drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 551,
   "id": "4f4bac0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "found_ORFs.to_csv('send/peptides_found_in_ncORFs.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 553,
   "id": "7b01a082",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(110, 94, 94)"
      ]
     },
     "execution_count": 553,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "found_ORFs.peptide.nunique(), found_ORFs['Amino acid sequence'].nunique(), found_ORFs['global_coo'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 557,
   "id": "68452af6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Amino acid sequence</th>\n",
       "      <th>peptide</th>\n",
       "      <th>pos</th>\n",
       "      <th>tr_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Type_of_ORF</th>\n",
       "      <th>Global_Rank_within_ORF_type</th>\n",
       "      <th>global_coo</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>MEEMVMIMASVGFVWSSPGLMEVGVGGPVVGGMGLLQEDLISEFLF...</td>\n",
       "      <td>FLFQDFLRQA</td>\n",
       "      <td>43</td>\n",
       "      <td>ENST00000229390.7</td>\n",
       "      <td>SRSF9</td>\n",
       "      <td>393</td>\n",
       "      <td>566</td>\n",
       "      <td>nested</td>\n",
       "      <td>98</td>\n",
       "      <td>chr12:120464090-120464122+chr12:120465627-1204...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>MEEMVMIMASVGFVWSSPGLMEVGVGGPVVGGMGLLQEDLISEFLF...</td>\n",
       "      <td>SEFLFQDF</td>\n",
       "      <td>41</td>\n",
       "      <td>ENST00000229390.7</td>\n",
       "      <td>SRSF9</td>\n",
       "      <td>393</td>\n",
       "      <td>566</td>\n",
       "      <td>nested</td>\n",
       "      <td>98</td>\n",
       "      <td>chr12:120464090-120464122+chr12:120465627-1204...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>MGGGGVGPLCHELQSHHFQKGHHR</td>\n",
       "      <td>HELQSHHF</td>\n",
       "      <td>10</td>\n",
       "      <td>ENST00000320623.9</td>\n",
       "      <td>NQO1</td>\n",
       "      <td>614</td>\n",
       "      <td>688</td>\n",
       "      <td>nested</td>\n",
       "      <td>934</td>\n",
       "      <td>chr16:69718250-69718253+chr16:69718370-69718440</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>MDSWCSPSWRMGAFPMRKISLRPSEEPVMEKP</td>\n",
       "      <td>SEEPVMEKP</td>\n",
       "      <td>23</td>\n",
       "      <td>ENST00000394231.7</td>\n",
       "      <td>MIEN1</td>\n",
       "      <td>501</td>\n",
       "      <td>599</td>\n",
       "      <td>nested</td>\n",
       "      <td>93</td>\n",
       "      <td>chr17:39729563-39729605+chr17:39729685-39729740</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MAGELQGTQAPSLRRSLG</td>\n",
       "      <td>GELQGTQAP</td>\n",
       "      <td>2</td>\n",
       "      <td>ENST00000368521.9</td>\n",
       "      <td>C1ORF43</td>\n",
       "      <td>90</td>\n",
       "      <td>146</td>\n",
       "      <td>uorf</td>\n",
       "      <td>919</td>\n",
       "      <td>chr1:154220461-154220517</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>MRRSASAARLAPLRHAC</td>\n",
       "      <td>SASAARLAPL</td>\n",
       "      <td>3</td>\n",
       "      <td>ENST00000269593.4</td>\n",
       "      <td>IGFBP4</td>\n",
       "      <td>45</td>\n",
       "      <td>98</td>\n",
       "      <td>uorf</td>\n",
       "      <td>1193</td>\n",
       "      <td>chr17:40443505-40443558</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>MELRNCVWSQPSLDAQEDAGGTVSEAWCQSCADPLLAMEQQAQRPS...</td>\n",
       "      <td>MEQQAQRPSP</td>\n",
       "      <td>37</td>\n",
       "      <td>ENST00000394697.6</td>\n",
       "      <td>ADCY7</td>\n",
       "      <td>69</td>\n",
       "      <td>248</td>\n",
       "      <td>uorf</td>\n",
       "      <td>1137</td>\n",
       "      <td>chr16:50266584-50266587+chr16:50287912-50288087</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>MSENVHFWDA</td>\n",
       "      <td>SENVHFWDA</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000311177.6</td>\n",
       "      <td>HOXB9</td>\n",
       "      <td>78</td>\n",
       "      <td>110</td>\n",
       "      <td>ouorf</td>\n",
       "      <td>112</td>\n",
       "      <td>chr17:48626247-48626279</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>MWTELGGPEPEPKRDPEPEPEPAAECQHAALPGTPAA</td>\n",
       "      <td>AECQHAALP</td>\n",
       "      <td>23</td>\n",
       "      <td>ENST00000377698.3</td>\n",
       "      <td>ALDH1B1</td>\n",
       "      <td>72</td>\n",
       "      <td>185</td>\n",
       "      <td>ouorf</td>\n",
       "      <td>109</td>\n",
       "      <td>chr9:38392735-38392807+chr9:38395740-38395780</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>MPDSNSLEDQDVDQLEDICNESSDNTQCVY</td>\n",
       "      <td>SSDNTQCVY</td>\n",
       "      <td>21</td>\n",
       "      <td>ENST00000296805.7</td>\n",
       "      <td>GFM2</td>\n",
       "      <td>428</td>\n",
       "      <td>520</td>\n",
       "      <td>ouorf</td>\n",
       "      <td>108</td>\n",
       "      <td>chr5:74763681-74763766+chr5:74766938-74766944</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Amino acid sequence     peptide  pos  \\\n",
       "94  MEEMVMIMASVGFVWSSPGLMEVGVGGPVVGGMGLLQEDLISEFLF...  FLFQDFLRQA   43   \n",
       "95  MEEMVMIMASVGFVWSSPGLMEVGVGGPVVGGMGLLQEDLISEFLF...    SEFLFQDF   41   \n",
       "91                           MGGGGVGPLCHELQSHHFQKGHHR    HELQSHHF   10   \n",
       "93                   MDSWCSPSWRMGAFPMRKISLRPSEEPVMEKP   SEEPVMEKP   23   \n",
       "9                                  MAGELQGTQAPSLRRSLG   GELQGTQAP    2   \n",
       "..                                                ...         ...  ...   \n",
       "38                                  MRRSASAARLAPLRHAC  SASAARLAPL    3   \n",
       "36  MELRNCVWSQPSLDAQEDAGGTVSEAWCQSCADPLLAMEQQAQRPS...  MEQQAQRPSP   37   \n",
       "83                                         MSENVHFWDA   SENVHFWDA    1   \n",
       "82              MWTELGGPEPEPKRDPEPEPEPAAECQHAALPGTPAA   AECQHAALP   23   \n",
       "81                     MPDSNSLEDQDVDQLEDICNESSDNTQCVY   SSDNTQCVY   21   \n",
       "\n",
       "                tr_id     Gene Start Stop Type_of_ORF  \\\n",
       "94  ENST00000229390.7    SRSF9   393  566      nested   \n",
       "95  ENST00000229390.7    SRSF9   393  566      nested   \n",
       "91  ENST00000320623.9     NQO1   614  688      nested   \n",
       "93  ENST00000394231.7    MIEN1   501  599      nested   \n",
       "9   ENST00000368521.9  C1ORF43    90  146        uorf   \n",
       "..                ...      ...   ...  ...         ...   \n",
       "38  ENST00000269593.4   IGFBP4    45   98        uorf   \n",
       "36  ENST00000394697.6    ADCY7    69  248        uorf   \n",
       "83  ENST00000311177.6    HOXB9    78  110       ouorf   \n",
       "82  ENST00000377698.3  ALDH1B1    72  185       ouorf   \n",
       "81  ENST00000296805.7     GFM2   428  520       ouorf   \n",
       "\n",
       "   Global_Rank_within_ORF_type  \\\n",
       "94                          98   \n",
       "95                          98   \n",
       "91                         934   \n",
       "93                          93   \n",
       "9                          919   \n",
       "..                         ...   \n",
       "38                        1193   \n",
       "36                        1137   \n",
       "83                         112   \n",
       "82                         109   \n",
       "81                         108   \n",
       "\n",
       "                                           global_coo strand  \n",
       "94  chr12:120464090-120464122+chr12:120465627-1204...      -  \n",
       "95  chr12:120464090-120464122+chr12:120465627-1204...      -  \n",
       "91    chr16:69718250-69718253+chr16:69718370-69718440      -  \n",
       "93    chr17:39729563-39729605+chr17:39729685-39729740      -  \n",
       "9                            chr1:154220461-154220517      -  \n",
       "..                                                ...    ...  \n",
       "38                            chr17:40443505-40443558      +  \n",
       "36    chr16:50266584-50266587+chr16:50287912-50288087      +  \n",
       "83                            chr17:48626247-48626279      -  \n",
       "82      chr9:38392735-38392807+chr9:38395740-38395780      +  \n",
       "81      chr5:74763681-74763766+chr5:74766938-74766944      -  \n",
       "\n",
       "[110 rows x 11 columns]"
      ]
     },
     "execution_count": 557,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "found_ORFs.sort_values(by='Global_Rank_within_ORF_type', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffa1eac4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9510173",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "115e954b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c6cacb6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
